src/components/ColumnPicker/useColumns.ts

```diff
@@ -0,0 +1,164 @@
+import useStateFromProps from "@dashboard/hooks/useStateFromProps";
+import { addAtIndex, removeAtIndex } from "@dashboard/utils/lists";
+import { GridColumn } from "@glideapps/glide-data-grid";
+import difference from "lodash/difference";
+import uniqBy from "lodash/uniqBy";
+import React from "react";
+
+import { AvailableColumn } from "../Datagrid/types";
+
+export interface ColumnCategory {
+  name: string;
+  prefix: string;
+  availableNodes: AvailableColumn[] | undefined;
+  selectedNodes: AvailableColumn[] | undefined;
+  initialSearch?: string;
+  onSearch?: (query: string) => void;
+  hasNextPage: boolean;
+  hasPreviousPage: boolean;
+  onNextPage: (query: string) => void;
+  onPreviousPage: (query: string) => void;
+}
+
+export interface UseColumnsProps {
+  staticColumns: AvailableColumn[];
+  columnCategories?: ColumnCategory[];
+  selectedColumns: string[];
+  onSave: (columns: string[]) => void;
+  columnPickerSettings?: string[];
+  setDynamicColumnSettings?: (cols: string[]) => void;
+}
+
+export const useColumns = ({
+  staticColumns,
+  selectedColumns,
+  columnCategories,
+  onSave,
+  columnPickerSettings,
+  setDynamicColumnSettings,
+}: UseColumnsProps) => {
+  const [dynamicColumns, updateDynamicColumns] =
+    React.useState<AvailableColumn[]>(null);
+
+  React.useEffect(() => {
+    if (
+      dynamicColumns === null &&
+      columnCategories?.every(category => Array.isArray(category.selectedNodes))
+    ) {
+      updateDynamicColumns(
+        columnCategories
+          .flatMap(category => category.selectedNodes)
+          .sort(
+            (a, b) =>
+              columnPickerSettings.indexOf(a.id) -
+              columnPickerSettings.indexOf(b.id),
+          ),
```

Comment:
**issue**: this is very hard to understand of what you want to do here, you can wrap it into function and that function name can indicate the meaning. Even you can split this function into smaller pieces, so the other developer can see a clear transition